src/wallet/wallet2.cpp

```diff
@@ -3136,13 +3136,24 @@ bool wallet2::store_keys(const std::string& keys_file_name, const epee::wipeable
   crypto::chacha20(account_data.data(), account_data.size(), key, keys_file_data.iv, &cipher[0]);
   keys_file_data.account_data = cipher;
 
-  unlock_keys_file();
+  std::string tmp_file_name = keys_file_name + ".new";
   std::string buf;
   r = ::serialization::dump_binary(keys_file_data, buf);
-  r = r && epee::file_io_utils::save_string_to_file(keys_file_name, buf); //and never touch wallet_keys_file again, only read
-  CHECK_AND_ASSERT_MES(r, false, "failed to generate wallet keys file " << keys_file_name);
+  r = r && epee::file_io_utils::save_string_to_file(tmp_file_name, buf);
+  if (!r) {
+      MERROR("Failed to generate wallet keys file " << tmp_file_name);
+      return false;
+  }
+
+  unlock_keys_file();
+  std::error_code e = tools::replace_file(tmp_file_name, keys_file_name);
   lock_keys_file();
 
+  if (e) {
+      MERROR("Failed to update wallet keys file " << keys_file_name);
+      return false;
+  }
+
   return true;
```

Comment:
Doesn't seem to remove the temp file on success

Reply:
OK, on error :)

Reply:
Oh right, sorry. :) Fixed